// @langfuse/client/dist/index.mjs
// src/LangfuseClient.ts
import {
  LangfuseAPIClient as LangfuseAPIClient4,
  LANGFUSE_SDK_VERSION,
  getGlobalLogger as getGlobalLogger6,
  getEnv as getEnv2
} from "@langfuse/core";

// src/dataset/index.ts
var DatasetManager = class {
  /**
   * Creates a new DatasetManager instance.
   *
   * @param params - Configuration object containing the API client
   * @internal
   */
  constructor(params) {
    this.langfuseClient = params.langfuseClient;
  }
  /**
   * Retrieves a dataset by name with all its items and experiment functionality.
   *
   * This method fetches a dataset and all its associated items, with support
   * for automatic pagination to handle large datasets efficiently. The returned
   * dataset object includes enhanced functionality for linking items to traces
   * and running experiments directly on the dataset.
   *
   * @param name - The name of the dataset to retrieve
   * @param options - Optional configuration for data fetching
   * @param options.fetchItemsPageSize - Number of items to fetch per page (default: 50)
   * @returns Promise resolving to enhanced dataset with items, linking, and experiment capabilities
   *
   * @example Basic dataset retrieval
   * ```typescript
   * const dataset = await langfuse.dataset.get("my-evaluation-dataset");
   * console.log(`Dataset ${dataset.name} has ${dataset.items.length} items`);
   *
   * // Access dataset properties
   * console.log(dataset.description);
   * console.log(dataset.metadata);
   * ```
   *
   * @example Working with dataset items
   * ```typescript
   * const dataset = await langfuse.dataset.get("qa-dataset");
   *
   * for (const item of dataset.items) {
   *   console.log("Question:", item.input);
   *   console.log("Expected Answer:", item.expectedOutput);
   *
   *   // Each item has a link function for connecting to traces
   *   // await item.link(span, "experiment-name");
   * }
   * ```
   *
   * @example Running experiments on datasets
   * ```typescript
   * const dataset = await langfuse.dataset.get("benchmark-dataset");
   *
   * const result = await dataset.runExperiment({
   *   name: "GPT-4 Benchmark",
   *   runName: "GPT-4 Benchmark v1.2", // optional exact run name
   *   description: "Evaluating GPT-4 on our benchmark tasks",
   *   task: async ({ input }) => {
   *     const response = await openai.chat.completions.create({
   *       model: "gpt-4",
   *       messages: [{ role: "user", content: input }]
   *     });
   *     return response.choices[0].message.content;
   *   },
   *   evaluators: [
   *     async ({ output, expectedOutput }) => ({
   *       name: "exact_match",
   *       value: output === expectedOutput ? 1 : 0
   *     })
   *   ]
   * });
   *
   * console.log(await result.format());
   * ```
   *
   * @example Handling large datasets
   * ```typescript
   * // For very large datasets, use smaller page sizes
   * const largeDataset = await langfuse.dataset.get(
   *   "large-dataset",
   *   { fetchItemsPageSize: 100 }
   * );
   * ```
   *
   * @throws {Error} If the dataset does not exist or cannot be accessed
   * @see {@link FetchedDataset} for the complete return type specification
   * @see {@link RunExperimentOnDataset} for experiment execution details
   * @public
   * @since 4.0.0
   */
  async get(name, options) {
    var _a;
    const dataset = await this.langfuseClient.api.datasets.get(name);
    const items = [];
    let page = 1;
    while (true) {
      const itemsResponse = await this.langfuseClient.api.datasetItems.list({
        datasetName: name,
        limit: (_a = options == null ? void 0 : options.fetchItemsPageSize) != null ? _a : 50,
        page,
        ...(options == null ? void 0 : options.version) && { version: options.version }
      });
      items.push(...itemsResponse.data);
      if (itemsResponse.meta.totalPages <= page) {
        break;
      }
      page++;
    }
    const itemsWithLinkMethod = items.map((item) => ({
      ...item,
      link: this.createDatasetItemLinkFunction(item)
    }));
    const runExperiment = (params) => {
      return this.langfuseClient.experiment.run({
        data: items,
        datasetVersion: options == null ? void 0 : options.version,
        ...params
      });
    };
    const returnDataset = {
      ...dataset,
      items: itemsWithLinkMethod,
      version: options == null ? void 0 : options.version,
      runExperiment
    };
    return returnDataset;
  }
  /**
   * Creates a link function for a specific dataset item.
   *
   * @param item - The dataset item to create a link function for
   * @returns A function that can link the item to OpenTelemetry spans
   * @internal
   */
  createDatasetItemLinkFunction(item) {
    const linkFunction = async (obj, runName, runArgs) => {
      return await this.langfuseClient.api.datasetRunItems.create({
        runName,
        datasetItemId: item.id,
        traceId: obj.otelSpan.spanContext().traceId,
        runDescription: runArgs == null ? void 0 : runArgs.description,
        metadata: runArgs == null ? void 0 : runArgs.metadata
      });
    };
    return linkFunction;
  }
};

// src/experiment/ExperimentManager.ts
import {
  getGlobalLogger,
  propagateAttributes,
  serializeValue,
  createExperimentId,
  createExperimentItemId,
  LangfuseOtelSpanAttributes,
  LANGFUSE_SDK_EXPERIMENT_ENVIRONMENT
} from "@langfuse/core";
import { startActiveObservation } from "@langfuse/tracing";
import { ProxyTracerProvider, trace } from "@opentelemetry/api";
var ExperimentManager = class {
  /**
   * Creates a new ExperimentManager instance.
   *
   * @param params - Configuration object
   * @param params.langfuseClient - The Langfuse client instance for API communication
   * @internal
   */
  constructor(params) {
    this.langfuseClient = params.langfuseClient;
  }
  /**
   * Gets the global logger instance for experiment-related logging.
   *
   * @returns The global logger instance
   * @internal
   */
  get logger() {
    return getGlobalLogger();
  }
  /**
   * Executes an experiment by running a task on each data item and evaluating the results.
   *
   * This method orchestrates the complete experiment lifecycle:
   * 1. Executes the task function on each data item with proper tracing
   * 2. Runs item-level evaluators on each task output
   * 3. Executes run-level evaluators on the complete result set
   * 4. Links results to dataset runs (for Langfuse datasets)
   * 5. Stores all scores and traces in Langfuse
   *
   * @param config - The experiment configuration
   * @param config.name - Human-readable name for the experiment
   * @param config.runName - Optional exact name for the experiment run (defaults to name + timestamp)
   * @param config.description - Optional description of the experiment's purpose
   * @param config.metadata - Optional metadata to attach to the experiment run
   * @param config.data - Array of data items to process (ExperimentItem[] or DatasetItem[])
   * @param config.task - Function that processes each data item and returns output
   * @param config.evaluators - Optional array of functions to evaluate each item's output
   * @param config.runEvaluators - Optional array of functions to evaluate the entire run
   * @param config.maxConcurrency - Maximum number of concurrent task executions (default: 50)
   *
   * @returns Promise that resolves to experiment results including:
   *   - experimentId: Stable identifier for the experiment execution
   *   - runName: The experiment run name (either provided or generated)
   *   - itemResults: Results for each processed data item
   *   - runEvaluations: Results from run-level evaluators
   *   - datasetRunId: ID of the dataset run (if using Langfuse datasets)
   *   - format: Function to format results for display
   *
   * @throws {Error} When task execution fails and cannot be handled gracefully
   * @throws {Error} When required evaluators fail critically
   *
   * @example Simple experiment
   * ```typescript
   * const result = await langfuse.experiment.run({
   *   name: "Translation Quality Test",
   *   data: [
   *     { input: "Hello world", expectedOutput: "Hola mundo" },
   *     { input: "Good morning", expectedOutput: "Buenos días" }
   *   ],
   *   task: async ({ input }) => translateText(input, 'es'),
   *   evaluators: [
   *     async ({ output, expectedOutput }) => ({
   *       name: "bleu_score",
   *       value: calculateBleuScore(output, expectedOutput)
   *     })
   *   ]
   * });
   * ```
   *
   * @example Experiment with concurrency control
   * ```typescript
   * const result = await langfuse.experiment.run({
   *   name: "Large Scale Evaluation",
   *   data: largeBatchOfItems,
   *   task: expensiveModelCall,
   *   maxConcurrency: 5, // Process max 5 items simultaneously
   *   evaluators: [myEvaluator],
   *   runEvaluators: [
   *     async ({ itemResults }) => ({
   *       name: "average_score",
   *       value: itemResults.reduce((acc, r) => acc + r.evaluations[0].value, 0) / itemResults.length
   *     })
   *   ]
   * });
   * ```
   *
   * @see {@link ExperimentParams} for detailed parameter documentation
   * @see {@link ExperimentResult} for detailed return value documentation
   * @see {@link Evaluator} for evaluator function specifications
   * @see {@link RunEvaluator} for run evaluator function specifications
   *
   * @public
   */
  async run(config) {
    var _a;
    const {
      data,
      evaluators,
      task,
      name,
      runName: providedRunName,
      description,
      metadata,
      maxConcurrency: batchSize = 50,
      runEvaluators
    } = config;
    const runName = this.createExperimentRunName({
      name,
      runName: providedRunName
    });
    const fallbackExperimentId = await createExperimentId();
    if (!this.isOtelRegistered()) {
      this.logger.warn(
        "OpenTelemetry has not been set up. Traces will not be sent to Langfuse.See our docs on how to set up OpenTelemetry: https://langfuse.com/docs/observability/sdk/typescript/setup#tracing-setup"
      );
    }
    const itemResults = [];
    for (let i = 0; i < data.length; i += batchSize) {
      const batch = data.slice(i, i + batchSize);
      const promises = batch.map(async (item) => {
        return this.runItem({
          item,
          evaluators,
          task,
          experimentName: name,
          experimentRunName: runName,
          experimentDescription: description,
          experimentMetadata: metadata,
          fallbackExperimentId,
          datasetVersion: config.datasetVersion
        });
      });
      const settledResults = await Promise.allSettled(promises);
      const results = settledResults.reduce(
        (acc, settledResult) => {
          if (settledResult.status === "fulfilled") {
            acc.push(settledResult.value);
          } else {
            const errorMessage = settledResult.reason instanceof Error ? settledResult.reason.message : String(settledResult.reason);
            this.logger.error(
              `Task failed with error: ${errorMessage}. Skipping item.`
            );
          }
          return acc;
        },
        []
      );
      itemResults.push(...results);
    }
    const datasetRunId = (_a = itemResults.find(
      (item) => item.datasetRunId
    )) == null ? void 0 : _a.datasetRunId;
    const experimentId = datasetRunId || fallbackExperimentId;
    let datasetRunUrl = void 0;
    if (datasetRunId && data.length > 0 && "datasetId" in data[0]) {
      const datasetId = data[0].datasetId;
      const projectUrl = (await this.langfuseClient.getTraceUrl("mock")).split(
        "/traces"
      )[0];
      datasetRunUrl = `${projectUrl}/datasets/${datasetId}/runs/${datasetRunId}`;
    }
    let runEvaluations = [];
    if (runEvaluators && (runEvaluators == null ? void 0 : runEvaluators.length) > 0) {
      const promises = runEvaluators.map(async (runEvaluator) => {
        return runEvaluator({ itemResults }).then((result) => {
          return Array.isArray(result) ? result : [result];
        }).catch((err) => {
          this.logger.error("Run evaluator failed with error ", err);
          throw err;
        });
      });
      runEvaluations = (await Promise.allSettled(promises)).reduce(
        (acc, settledPromise) => {
          if (settledPromise.status === "fulfilled") {
            acc.push(...settledPromise.value);
          }
          return acc;
        },
        []
      );
      if (datasetRunId) {
        runEvaluations.forEach(
          (runEval) => this.langfuseClient.score.create({ datasetRunId, ...runEval })
        );
      }
    }
    await this.langfuseClient.score.flush();
    return {
      experimentId,
      runName,
      itemResults,
      datasetRunId,
      datasetRunUrl,
      runEvaluations,
      format: async (options) => {
        var _a2;
        return await this.prettyPrintResults({
          datasetRunUrl,
          itemResults,
          originalData: data,
          runEvaluations,
          name: config.name,
          runName,
          description: config.description,
          includeItemResults: (_a2 = options == null ? void 0 : options.includeItemResults) != null ? _a2 : false
        });
      }
    };
  }
  /**
   * Executes the task and evaluators for a single data item.
   *
   * This method handles the complete processing pipeline for one data item:
   * 1. Executes the task within a traced observation span
   * 2. Links the result to a dataset run (if applicable)
   * 3. Runs all item-level evaluators on the output
   * 4. Stores evaluation scores in Langfuse
   * 5. Handles errors gracefully by continuing with remaining evaluators
   *
   * @param params - Parameters for item execution
   * @param params.experimentName - Name of the parent experiment
   * @param params.experimentRunName - Run name for the parent experiment
   * @param params.experimentDescription - Description of the parent experiment
   * @param params.experimentMetadata - Metadata for the parent experiment
   * @param params.item - The data item to process
   * @param params.task - The task function to execute
   * @param params.evaluators - Optional evaluators to run on the output
   *
   * @returns Promise resolving to the item result with output, evaluations, and trace info
   *
   * @throws {Error} When task execution fails (propagated from task function)
   *
   * @internal
   */
  async runItem(params) {
    const { item, evaluators = [], task, experimentMetadata } = params;
    const { output, traceId, observationId, datasetRunId } = await startActiveObservation("experiment-item-run", async (span) => {
      const input = item.input;
      const expectedOutput = item.expectedOutput;
      const itemMetadata = item.metadata;
      const datasetId = "datasetId" in item ? item.datasetId : void 0;
      const datasetItemId = "id" in item ? item.id : void 0;
      const traceId2 = span.traceId;
      const observationId2 = span.id;
      if (input === void 0) {
        throw new Error("Experiment item is missing input. Skipping item.");
      }
      let datasetRunId2 = void 0;
      if (datasetItemId) {
        try {
          const result = await this.langfuseClient.api.datasetRunItems.create(
            {
              runName: params.experimentRunName,
              runDescription: params.experimentDescription,
              metadata: params.experimentMetadata,
              datasetItemId,
              traceId: traceId2,
              observationId: observationId2,
              ...params.datasetVersion && {
                datasetVersion: params.datasetVersion
              }
            }
          );
          datasetRunId2 = result.datasetRunId;
        } catch (err) {
          this.logger.error("Linking dataset run item failed", err);
        }
      }
      const experimentItemId = datasetItemId || await createExperimentItemId(input);
      const experimentId = datasetRunId2 || params.fallbackExperimentId;
      const rootSpanAttributes = {
        [LangfuseOtelSpanAttributes.ENVIRONMENT]: LANGFUSE_SDK_EXPERIMENT_ENVIRONMENT
      };
      if (params.experimentDescription) {
        rootSpanAttributes[LangfuseOtelSpanAttributes.EXPERIMENT_DESCRIPTION] = params.experimentDescription;
      }
      if (expectedOutput !== void 0) {
        const serialized = serializeValue(expectedOutput);
        if (serialized) {
          rootSpanAttributes[LangfuseOtelSpanAttributes.EXPERIMENT_ITEM_EXPECTED_OUTPUT] = serialized;
        }
      }
      span.otelSpan.setAttributes(rootSpanAttributes);
      const output2 = await propagateAttributes(
        {
          _internalExperiment: {
            experimentId,
            experimentName: params.experimentRunName,
            experimentMetadata: serializeValue(experimentMetadata),
            experimentDatasetId: datasetId,
            experimentItemId,
            experimentItemMetadata: serializeValue(itemMetadata),
            experimentItemRootObservationId: span.id
          }
        },
        async () => await task(item)
      );
      span.update({
        input,
        output: output2,
        metadata: {
          experiment_name: params.experimentName,
          experiment_run_name: params.experimentRunName,
          ...experimentMetadata,
          ...itemMetadata != null ? itemMetadata : {},
          ...datasetId && datasetItemId ? {
            dataset_id: datasetId,
            dataset_item_id: datasetItemId
          } : {}
        }
      });
      return {
        output: output2,
        traceId: traceId2,
        observationId: observationId2,
        datasetRunId: datasetRunId2
      };
    });
    const evalPromises = evaluators.map(
      async (evaluator) => {
        const params2 = {
          input: item.input,
          expectedOutput: item.expectedOutput,
          output,
          metadata: item.metadata
        };
        return evaluator(params2).then((result) => {
          return Array.isArray(result) ? result : [result];
        }).catch((err) => {
          this.logger.error(
            `Evaluator '${evaluator.name}' failed for params 

${JSON.stringify(params2)}

 with error: ${err}`
          );
          throw err;
        });
      }
    );
    const evals = (await Promise.allSettled(evalPromises)).reduce(
      (acc, promiseResult) => {
        if (promiseResult.status === "fulfilled") {
          acc.push(...promiseResult.value.flat());
        }
        return acc;
      },
      []
    );
    for (const ev of evals) {
      this.langfuseClient.score.create({
        traceId,
        observationId,
        ...ev
      });
    }
    return {
      output,
      evaluations: evals,
      traceId,
      datasetRunId,
      item
    };
  }
  /**
   * Formats experiment results into a human-readable string representation.
   *
   * Creates a comprehensive, nicely formatted summary of the experiment including:
   * - Individual item results with inputs, outputs, expected values, and scores
   * - Dataset item and trace links (when available)
   * - Experiment overview with aggregate statistics
   * - Average scores across all evaluations
   * - Run-level evaluation results
   * - Links to dataset runs in the Langfuse UI
   *
   * @param params - Formatting parameters
   * @param params.datasetRunUrl - Optional URL to the dataset run in Langfuse UI
   * @param params.itemResults - Results from processing each data item
   * @param params.originalData - The original input data items
   * @param params.runEvaluations - Results from run-level evaluators
   * @param params.name - Name of the experiment
   * @param params.description - Optional description of the experiment
   * @param params.includeItemResults - Whether to include individual item details (default: false)
   *
   * @returns Promise resolving to formatted string representation
   *
   * @example Output format
   * ```
   * 1. Item 1:
   *    Input:    What is the capital of France?
   *    Expected: Paris
   *    Actual:   Paris
   *    Scores:
   *      • exact_match: 1.000
   *      • similarity: 0.95
   *        💭 Very close match with expected output
   *
   *    Dataset Item:
   *    https://cloud.langfuse.com/project/123/datasets/456/items/789
   *
   *    Trace:
   *    https://cloud.langfuse.com/project/123/traces/abc123
   *
   * ──────────────────────────────────────────────────
   * 📊 Translation Quality Test - Testing model accuracy
   * 2 items
   * Evaluations:
   *   • exact_match
   *   • similarity
   *
   * Average Scores:
   *   • exact_match: 0.850
   *   • similarity: 0.923
   *
   * Run Evaluations:
   *   • overall_quality: 0.887
   *     💭 Good performance with room for improvement
   *
   * 🔗 Dataset Run:
   *    https://cloud.langfuse.com/project/123/datasets/456/runs/def456
   * ```
   *
   * @internal
   */
  async prettyPrintResults(params) {
    var _a, _b;
    const {
      itemResults,
      originalData,
      runEvaluations,
      name,
      runName,
      description,
      includeItemResults = false
    } = params;
    if (itemResults.length === 0) {
      return "No experiment results to display.";
    }
    let output = "";
    if (includeItemResults) {
      for (let index = 0; index < itemResults.length; index++) {
        const result = itemResults[index];
        const originalItem = originalData[index];
        output += `
${index + 1}. Item ${index + 1}:
`;
        if ((originalItem == null ? void 0 : originalItem.input) !== void 0) {
          output += `   Input:    ${this.formatValue(originalItem.input)}
`;
        }
        const expectedOutput = (_b = (_a = originalItem == null ? void 0 : originalItem.expectedOutput) != null ? _a : result.expectedOutput) != null ? _b : null;
        output += `   Expected: ${expectedOutput !== null ? this.formatValue(expectedOutput) : "null"}
`;
        output += `   Actual:   ${this.formatValue(result.output)}
`;
        if (result.evaluations.length > 0) {
          output += `   Scores:
`;
          result.evaluations.forEach((evaluation) => {
            const score = typeof evaluation.value === "number" ? evaluation.value.toFixed(3) : evaluation.value;
            output += `     \u2022 ${evaluation.name}: ${score}`;
            if (evaluation.comment) {
              output += `
       \u{1F4AD} ${evaluation.comment}`;
            }
            output += "\n";
          });
        }
        if (originalItem && "id" in originalItem && "datasetId" in originalItem) {
          const projectUrl = (await this.langfuseClient.getTraceUrl("mock")).split("/traces")[0];
          const datasetItemUrl = `${projectUrl}/datasets/${originalItem.datasetId}/items/${originalItem.id}`;
          output += `
   Dataset Item:
   ${datasetItemUrl}
`;
        }
        if (result.traceId) {
          const traceUrl = await this.langfuseClient.getTraceUrl(
            result.traceId
          );
          output += `
   Trace:
   ${traceUrl}
`;
        }
      }
    } else {
      output += `Individual Results: Hidden (${itemResults.length} items)
`;
      output += "\u{1F4A1} Call format({ includeItemResults: true }) to view them\n";
    }
    const totalItems = itemResults.length;
    const evaluationNames = new Set(
      itemResults.flatMap((r) => r.evaluations.map((e) => e.name))
    );
    output += `
${"\u2500".repeat(50)}
`;
    output += `\u{1F9EA} Experiment: ${name}`;
    output += `
\u{1F4CB} Run name: ${runName}`;
    if (description) {
      output += ` - ${description}`;
    }
    output += `
${totalItems} items`;
    if (evaluationNames.size > 0) {
      output += `
Evaluations:`;
      Array.from(evaluationNames).forEach((evalName) => {
        output += `
  \u2022 ${evalName}`;
      });
      output += "\n";
    }
    if (evaluationNames.size > 0) {
      output += `
Average Scores:`;
      for (const evalName of evaluationNames) {
        const scores = itemResults.flatMap((r) => r.evaluations).filter((e) => e.name === evalName && typeof e.value === "number").map((e) => e.value);
        if (scores.length > 0) {
          const avg = scores.reduce((a, b) => a + b, 0) / scores.length;
          output += `
  \u2022 ${evalName}: ${avg.toFixed(3)}`;
        }
      }
      output += "\n";
    }
    if (runEvaluations.length > 0) {
      output += `
Run Evaluations:`;
      runEvaluations.forEach((runEval) => {
        const score = typeof runEval.value === "number" ? runEval.value.toFixed(3) : runEval.value;
        output += `
  \u2022 ${runEval.name}: ${score}`;
        if (runEval.comment) {
          output += `
    \u{1F4AD} ${runEval.comment}`;
        }
      });
      output += "\n";
    }
    if (params.datasetRunUrl) {
      output += `
\u{1F517} Dataset Run:
   ${params.datasetRunUrl}`;
    }
    return output;
  }
  /**
   * Formats a value for display in pretty-printed output.
   *
   * Handles different value types appropriately:
   * - Strings: Truncates long strings to 50 characters with "..."
   * - Objects/Arrays: Converts to JSON string representation
   * - Primitives: Uses toString() representation
   *
   * @param value - The value to format
   * @returns Formatted string representation suitable for display
   *
   * @internal
   */
  formatValue(value) {
    if (typeof value === "string") {
      return value.length > 50 ? `${value.substring(0, 47)}...` : value;
    }
    return JSON.stringify(value);
  }
  isOtelRegistered() {
    let tracerProvider = trace.getTracerProvider();
    if (tracerProvider instanceof ProxyTracerProvider) {
      tracerProvider = tracerProvider.getDelegate();
    }
    return tracerProvider.constructor.name !== "NoopTracerProvider";
  }
  /**
   * Creates an experiment run name based on provided parameters.
   *
   * If runName is provided, returns it directly. Otherwise, generates
   * a name by combining the experiment name with an ISO timestamp.
   *
   * @param params - Parameters for run name creation
   * @param params.name - The experiment name
   * @param params.runName - Optional provided run name
   * @returns The final run name to use
   *
   * @internal
   */
  createExperimentRunName(params) {
    if (params.runName) {
      return params.runName;
    }
    const isoTimestamp = (/* @__PURE__ */ new Date()).toISOString();
    return `${params.name} - ${isoTimestamp}`;
  }
};

// src/media/index.ts
import {
  getGlobalLogger as getGlobalLogger2,
  bytesToBase64
} from "@langfuse/core";
var MediaManager = class _MediaManager {
  /**
   * Creates a new MediaManager instance.
   *
   * @param params - Configuration object containing the API client
   * @internal
   */
  constructor(params) {
    this.apiClient = params.apiClient;
  }
  /**
   * Replaces media reference strings in an object with base64 data URIs.
   *
   * This method recursively traverses an object looking for media reference strings
   * in the format "@@@langfuseMedia:...@@@". When found, it fetches the actual media
   * content from Langfuse and replaces the reference string with a base64 data URI.
   *
   * If fetching media content fails for a reference string, a warning is logged
   * and the reference string is left unchanged.
   *
   * @param params - Configuration object
   * @returns A deep copy of the input object with media references resolved
   *
   * @example
   * ```typescript
   * const obj = {
   *   image: "@@@langfuseMedia:type=image/jpeg|id=123|source=bytes@@@",
   *   nested: {
   *     pdf: "@@@langfuseMedia:type=application/pdf|id=456|source=bytes@@@"
   *   }
   * };
   *
   * const result = await langfuse.media.resolveReferences({
   *   obj,
   *   resolveWith: "base64DataUri"
   * });
   *
   * // Result:
   * // {
   * //   image: "data:image/jpeg;base64,/9j/4AAQSkZJRg...",
   * //   nested: {
   * //     pdf: "data:application/pdf;base64,JVBERi0xLjcK..."
   * //   }
   * // }
   * ```
   */
  async resolveReferences(params) {
    const { obj, maxDepth = 10 } = params;
    const traverse = async (obj2, depth) => {
      if (depth > maxDepth) {
        return obj2;
      }
      if (typeof obj2 === "string") {
        const regex = /@@@langfuseMedia:.+?@@@/g;
        const referenceStringMatches = obj2.match(regex);
        if (!referenceStringMatches) {
          return obj2;
        }
        let result = obj2;
        const referenceStringToMediaContentMap = /* @__PURE__ */ new Map();
        await Promise.all(
          referenceStringMatches.map(async (referenceString) => {
            try {
              const parsedMediaReference = _MediaManager.parseReferenceString(referenceString);
              const mediaData = await this.apiClient.media.get(
                parsedMediaReference.mediaId
              );
              const mediaContent = await fetch(mediaData.url, {
                method: "GET",
                headers: {}
              });
              if (mediaContent.status !== 200) {
                throw new Error("Failed to fetch media content");
              }
              const uint8Content = new Uint8Array(
                await mediaContent.arrayBuffer()
              );
              const base64MediaContent = bytesToBase64(uint8Content);
              const base64DataUri = `data:${mediaData.contentType};base64,${base64MediaContent}`;
              referenceStringToMediaContentMap.set(
                referenceString,
                base64DataUri
              );
            } catch (error) {
              getGlobalLogger2().warn(
                "Error fetching media content for reference string",
                referenceString,
                error
              );
            }
          })
        );
        for (const [
          referenceString,
          base64MediaContent
        ] of referenceStringToMediaContentMap.entries()) {
          result = result.replaceAll(referenceString, base64MediaContent);
        }
        return result;
      }
      if (Array.isArray(obj2)) {
        return Promise.all(
          obj2.map(async (item) => await traverse(item, depth + 1))
        );
      }
      if (typeof obj2 === "object" && obj2 !== null) {
        return Object.fromEntries(
          await Promise.all(
            Object.entries(obj2).map(async ([key, value]) => [
              key,
              await traverse(value, depth + 1)
            ])
          )
        );
      }
      return obj2;
    };
    return traverse(obj, 0);
  }
  /**
   * Parses a media reference string into a ParsedMediaReference.
   *
   * Example reference string:
   *     "@@@langfuseMedia:type=image/jpeg|id=some-uuid|source=base64DataUri@@@"
   *
   * @param referenceString - The reference string to parse.
   * @returns An object with the mediaId, source, and contentType.
   *
   * @throws Error if the reference string is invalid or missing required fields.
   */
  static parseReferenceString(referenceString) {
    const prefix = "@@@langfuseMedia:";
    const suffix = "@@@";
    if (!referenceString.startsWith(prefix)) {
      throw new Error(
        "Reference string does not start with '@@@langfuseMedia:type='"
      );
    }
    if (!referenceString.endsWith(suffix)) {
      throw new Error("Reference string does not end with '@@@'");
    }
    const content = referenceString.slice(prefix.length, -suffix.length);
    const pairs = content.split("|");
    const parsedData = {};
    for (const pair of pairs) {
      const [key, value] = pair.split("=", 2);
      parsedData[key] = value;
    }
    if (!("type" in parsedData && "id" in parsedData && "source" in parsedData)) {
      throw new Error("Missing required fields in reference string");
    }
    return {
      mediaId: parsedData["id"],
      source: parsedData["source"],
      contentType: parsedData["type"]
    };
  }
};

// src/prompt/promptManager.ts
import {
  getGlobalLogger as getGlobalLogger4
} from "@langfuse/core";

// src/prompt/promptCache.ts
import { getGlobalLogger as getGlobalLogger3 } from "@langfuse/core";
var DEFAULT_PROMPT_CACHE_TTL_SECONDS = 60;
var LangfusePromptCacheItem = class {
  constructor(value, ttlSeconds) {
    this.value = value;
    this._expiry = Date.now() + ttlSeconds * 1e3;
  }
  get isExpired() {
    return Date.now() > this._expiry;
  }
};
var LangfusePromptCache = class {
  constructor() {
    this._cache = /* @__PURE__ */ new Map();
    this._defaultTtlSeconds = DEFAULT_PROMPT_CACHE_TTL_SECONDS;
    this._refreshingKeys = /* @__PURE__ */ new Map();
  }
  getIncludingExpired(key) {
    var _a;
    return (_a = this._cache.get(key)) != null ? _a : null;
  }
  createKey(params) {
    const { name, version, label } = params;
    const parts = [name];
    if (version !== void 0) {
      parts.push("version:" + version.toString());
    } else if (label !== void 0) {
      parts.push("label:" + label);
    } else {
      parts.push("label:production");
    }
    return parts.join("-");
  }
  set(key, value, ttlSeconds) {
    const effectiveTtlSeconds = ttlSeconds != null ? ttlSeconds : this._defaultTtlSeconds;
    this._cache.set(
      key,
      new LangfusePromptCacheItem(value, effectiveTtlSeconds)
    );
  }
  addRefreshingPromise(key, promise) {
    this._refreshingKeys.set(key, promise);
    promise.then(() => {
      this._refreshingKeys.delete(key);
    }).catch(() => {
      this._refreshingKeys.delete(key);
    });
  }
  isRefreshing(key) {
    return this._refreshingKeys.has(key);
  }
  invalidate(promptName) {
    getGlobalLogger3().debug(
      "Invalidating cache keys for",
      promptName,
      this._cache.keys()
    );
    for (const key of this._cache.keys()) {
      if (key.startsWith(promptName)) {
        this._cache.delete(key);
      }
    }
  }
};

// src/prompt/promptClients.ts
import mustache from "mustache";

// src/prompt/types.ts
var ChatMessageType = /* @__PURE__ */ ((ChatMessageType2) => {
  ChatMessageType2["ChatMessage"] = "chatmessage";
  ChatMessageType2["Placeholder"] = "placeholder";
  return ChatMessageType2;
})(ChatMessageType || {});

// src/prompt/promptClients.ts
mustache.escape = function(text) {
  return text;
};
var BasePromptClient = class {
  /**
   * Creates a new BasePromptClient instance.
   *
   * @param prompt - The base prompt data
   * @param isFallback - Whether this is fallback content
   * @param type - The prompt type
   * @internal
   */
  constructor(prompt, isFallback = false, type) {
    this.name = prompt.name;
    this.version = prompt.version;
    this.config = prompt.config;
    this.labels = prompt.labels;
    this.tags = prompt.tags;
    this.isFallback = isFallback;
    this.type = type;
    this.commitMessage = prompt.commitMessage;
  }
  _transformToLangchainVariables(content) {
    const jsonEscapedContent = this.escapeJsonForLangchain(content);
    return jsonEscapedContent.replace(/\{\{(\w+)\}\}/g, "{$1}");
  }
  /**
   * Escapes every curly brace that is part of a JSON object by doubling it.
   *
   * A curly brace is considered “JSON-related” when, after skipping any immediate
   * whitespace, the next non-whitespace character is a single (') or double (") quote.
   *
   * Braces that are already doubled (e.g. `{{variable}}` placeholders) are left untouched.
   *
   * @param text - Input string that may contain JSON snippets.
   * @returns The string with JSON-related braces doubled.
   */
  escapeJsonForLangchain(text) {
    var _a;
    const out = [];
    const stack = [];
    let i = 0;
    const n = text.length;
    while (i < n) {
      const ch = text[i];
      if (ch === "{") {
        if (i + 1 < n && text[i + 1] === "{") {
          out.push("{{");
          i += 2;
          continue;
        }
        let j = i + 1;
        while (j < n && /\s/.test(text[j])) {
          j++;
        }
        const isJson = j < n && (text[j] === "'" || text[j] === '"');
        out.push(isJson ? "{{" : "{");
        stack.push(isJson);
        i += 1;
        continue;
      }
      if (ch === "}") {
        if (i + 1 < n && text[i + 1] === "}") {
          out.push("}}");
          i += 2;
          continue;
        }
        const isJson = (_a = stack.pop()) != null ? _a : false;
        out.push(isJson ? "}}" : "}");
        i += 1;
        continue;
      }
      out.push(ch);
      i += 1;
    }
    return out.join("");
  }
};
var TextPromptClient = class extends BasePromptClient {
  /**
   * Creates a new TextPromptClient instance.
   *
   * @param prompt - The text prompt data
   * @param isFallback - Whether this is fallback content
   */
  constructor(prompt, isFallback = false) {
    super(prompt, isFallback, "text");
    this.promptResponse = prompt;
    this.prompt = prompt.prompt;
  }
  /**
   * Compiles the text prompt by substituting variables.
   *
   * Uses Mustache templating to replace {{variable}} placeholders with provided values.
   *
   * @param variables - Key-value pairs for variable substitution
   * @param _placeholders - Ignored for text prompts
   * @returns The compiled text with variables substituted
   *
   * @example
   * ```typescript
   * const prompt = await langfuse.prompt.get("greeting", { type: "text" });
   * const compiled = prompt.compile({ name: "Alice" });
   * // If prompt is "Hello {{name}}!", result is "Hello Alice!"
   * ```
   */
  compile(variables, _placeholders) {
    return mustache.render(this.promptResponse.prompt, variables != null ? variables : {});
  }
  /**
   * Converts the prompt to LangChain PromptTemplate format.
   *
   * Transforms Mustache-style {{variable}} syntax to LangChain's {variable} format.
   *
   * @param _options - Ignored for text prompts
   * @returns The prompt string compatible with LangChain PromptTemplate
   *
   * @example
   * ```typescript
   * const prompt = await langfuse.prompt.get("greeting", { type: "text" });
   * const langchainFormat = prompt.getLangchainPrompt();
   * // Transforms "Hello {{name}}!" to "Hello {name}!"
   * ```
   */
  getLangchainPrompt(_options) {
    return this._transformToLangchainVariables(this.prompt);
  }
  toJSON() {
    return JSON.stringify({
      name: this.name,
      prompt: this.prompt,
      version: this.version,
      isFallback: this.isFallback,
      tags: this.tags,
      labels: this.labels,
      type: this.type,
      config: this.config
    });
  }
};
var ChatPromptClient = class _ChatPromptClient extends BasePromptClient {
  /**
   * Creates a new ChatPromptClient instance.
   *
   * @param prompt - The chat prompt data
   * @param isFallback - Whether this is fallback content
   */
  constructor(prompt, isFallback = false) {
    const normalizedPrompt = _ChatPromptClient.normalizePrompt(prompt.prompt);
    const typedPrompt = {
      ...prompt,
      prompt: normalizedPrompt
    };
    super(typedPrompt, isFallback, "chat");
    this.promptResponse = typedPrompt;
    this.prompt = normalizedPrompt;
  }
  static normalizePrompt(prompt) {
    return prompt.map((item) => {
      if ("type" in item) {
        return item;
      } else {
        return {
          type: "chatmessage" /* ChatMessage */,
          ...item
        };
      }
    });
  }
  /**
   * Compiles the chat prompt by replacing placeholders and variables.
   *
   * First resolves placeholders with provided values, then applies variable substitution
   * to message content using Mustache templating. Unresolved placeholders remain
   * as placeholder objects in the output.
   *
   * @param variables - Key-value pairs for Mustache variable substitution in message content
   * @param placeholders - Key-value pairs where keys are placeholder names and values are ChatMessage arrays
   * @returns Array of ChatMessage objects and unresolved placeholder objects
   *
   * @example
   * ```typescript
   * const prompt = await langfuse.prompt.get("conversation", { type: "chat" });
   * const compiled = prompt.compile(
   *   { user_name: "Alice" },
   *   { examples: [{ role: "user", content: "Hello" }, { role: "assistant", content: "Hi!" }] }
   * );
   * ```
   */
  compile(variables, placeholders) {
    const messagesWithPlaceholdersReplaced = [];
    const placeholderValues = placeholders != null ? placeholders : {};
    for (const item of this.prompt) {
      if ("type" in item && item.type === "placeholder" /* Placeholder */) {
        const placeholderValue = placeholderValues[item.name];
        if (Array.isArray(placeholderValue) && placeholderValue.length > 0 && placeholderValue.every(
          (msg) => typeof msg === "object" && "role" in msg && "content" in msg
        )) {
          messagesWithPlaceholdersReplaced.push(
            ...placeholderValue
          );
        } else if (Array.isArray(placeholderValue) && placeholderValue.length === 0) {
        } else if (placeholderValue !== void 0) {
          messagesWithPlaceholdersReplaced.push(
            JSON.stringify(placeholderValue)
          );
        } else {
          messagesWithPlaceholdersReplaced.push(
            item
          );
        }
      } else if ("role" in item && "content" in item && item.type === "chatmessage" /* ChatMessage */) {
        messagesWithPlaceholdersReplaced.push({
          role: item.role,
          content: item.content
        });
      }
    }
    return messagesWithPlaceholdersReplaced.map((item) => {
      if (typeof item === "object" && item !== null && "role" in item && "content" in item && typeof item.content === "string") {
        return {
          ...item,
          content: mustache.render(item.content, variables != null ? variables : {})
        };
      } else {
        return item;
      }
    });
  }
  /**
   * Converts the prompt to LangChain ChatPromptTemplate format.
   *
   * Resolves placeholders with provided values and converts unresolved ones
   * to LangChain MessagesPlaceholder objects. Transforms variables from
   * {{var}} to {var} format without rendering them.
   *
   * @param options - Configuration object
   * @param options.placeholders - Key-value pairs for placeholder resolution
   * @returns Array of ChatMessage objects and LangChain MessagesPlaceholder objects
   *
   * @example
   * ```typescript
   * const prompt = await langfuse.prompt.get("conversation", { type: "chat" });
   * const langchainFormat = prompt.getLangchainPrompt({
   *   placeholders: { examples: [{ role: "user", content: "Hello" }] }
   * });
   * ```
   */
  getLangchainPrompt(options) {
    var _a;
    const messagesWithPlaceholdersReplaced = [];
    const placeholderValues = (_a = options == null ? void 0 : options.placeholders) != null ? _a : {};
    for (const item of this.prompt) {
      if ("type" in item && item.type === "placeholder" /* Placeholder */) {
        const placeholderValue = placeholderValues[item.name];
        if (Array.isArray(placeholderValue) && placeholderValue.length > 0 && placeholderValue.every(
          (msg) => typeof msg === "object" && "role" in msg && "content" in msg
        )) {
          messagesWithPlaceholdersReplaced.push(
            ...placeholderValue.map((msg) => {
              return {
                role: msg.role,
                content: this._transformToLangchainVariables(msg.content)
              };
            })
          );
        } else if (Array.isArray(placeholderValue) && placeholderValue.length === 0) {
        } else if (placeholderValue !== void 0) {
          messagesWithPlaceholdersReplaced.push(
            JSON.stringify(placeholderValue)
          );
        } else {
          messagesWithPlaceholdersReplaced.push([
            "placeholder",
            `{${item.name}}`
          ]);
        }
      } else if ("role" in item && "content" in item && item.type === "chatmessage" /* ChatMessage */) {
        messagesWithPlaceholdersReplaced.push({
          role: item.role,
          content: this._transformToLangchainVariables(item.content)
        });
      }
    }
    return messagesWithPlaceholdersReplaced;
  }
  toJSON() {
    return JSON.stringify({
      name: this.name,
      prompt: this.promptResponse.prompt.map((item) => {
        if ("type" in item && item.type === "chatmessage" /* ChatMessage */) {
          const { type: _, ...messageWithoutType } = item;
          return messageWithoutType;
        }
        return item;
      }),
      version: this.version,
      isFallback: this.isFallback,
      tags: this.tags,
      labels: this.labels,
      type: this.type,
      config: this.config
    });
  }
};

// src/prompt/promptManager.ts
var PromptManager = class {
  /**
   * Creates a new PromptManager instance.
   *
   * @param params - Configuration object containing the API client
   * @internal
   */
  constructor(params) {
    const { apiClient } = params;
    this.apiClient = apiClient;
    this.cache = new LangfusePromptCache();
  }
  get logger() {
    return getGlobalLogger4();
  }
  /**
   * Creates a new prompt in Langfuse.
   *
   * Supports both text and chat prompts. Chat prompts can include placeholders
   * for dynamic content insertion.
   *
   * @param body - The prompt data to create
   * @returns Promise that resolves to the appropriate prompt client
   *
   * @example
   * ```typescript
   * // Create a text prompt
   * const textPrompt = await langfuse.prompt.create({
   *   name: "greeting",
   *   prompt: "Hello {{name}}!",
   *   type: "text"
   * });
   *
   * // Create a chat prompt
   * const chatPrompt = await langfuse.prompt.create({
   *   name: "conversation",
   *   type: "chat",
   *   prompt: [
   *     { role: "system", content: "You are a helpful assistant." },
   *     { role: "user", content: "{{user_message}}" }
   *   ]
   * });
   * ```
   */
  async create(body) {
    var _a;
    const requestBody = body.type === "chat" ? {
      ...body,
      prompt: body.prompt.map((item) => {
        if ("type" in item && item.type === "placeholder" /* Placeholder */) {
          return {
            type: "placeholder" /* Placeholder */,
            name: item.name
          };
        } else {
          return {
            ...item,
            type: "chatmessage" /* ChatMessage */
          };
        }
      })
    } : {
      ...body,
      type: (_a = body.type) != null ? _a : "text"
    };
    const promptResponse = await this.apiClient.prompts.create(requestBody);
    if (promptResponse.type === "chat") {
      return new ChatPromptClient(promptResponse);
    }
    return new TextPromptClient(promptResponse);
  }
  /**
   * Updates the labels of an existing prompt version.
   *
   * @param params - Update parameters
   * @param params.name - Name of the prompt to update
   * @param params.version - Version number of the prompt to update
   * @param params.newLabels - New labels to apply to the prompt version
   *
   * @returns Promise that resolves to the updated prompt
   *
   * @example
   * ```typescript
   * const updatedPrompt = await langfuse.prompt.update({
   *   name: "my-prompt",
   *   version: 1,
   *   newLabels: ["production", "v2"]
   * });
   * ```
   */
  async update(params) {
    const { name, version, newLabels } = params;
    const newPrompt = await this.apiClient.promptVersion.update(name, version, {
      newLabels
    });
    this.cache.invalidate(name);
    return newPrompt;
  }
  /**
   * Delete prompt versions. If neither version nor label is specified, all versions of the prompt are deleted.
   *
   * The Langfuse SDK prompt cache is invalidated for all cached versions with the specified name.
   *
   * @param name - Name of the prompt to delete
   * @param options - Optional deletion configuration
   * @param options.version - Optional version to delete. If specified, deletes only this specific version
   * @param options.label - Optional label to filter deletion. If specified, deletes all prompt versions with this label
   *
   * @returns Promise that resolves when deletion is complete
   *
   * @throws {LangfuseAPI.NotFoundError} If the prompt does not exist
   * @throws {LangfuseAPI.Error} If the API request fails
   *
   * @example
   * ```typescript
   * // Delete all versions of a prompt
   * await langfuse.prompt.delete("my-prompt");
   *
   * // Delete specific version
   * await langfuse.prompt.delete("my-prompt", { version: 2 });
   *
   * // Delete all versions with a specific label
   * await langfuse.prompt.delete("my-prompt", { label: "staging" });
   * ```
   */
  async delete(name, options) {
    await this.apiClient.prompts.delete(name, {
      version: options == null ? void 0 : options.version,
      label: options == null ? void 0 : options.label
    });
    this.cache.invalidate(name);
  }
  /**
   * Retrieves a prompt by name with intelligent caching.
   *
   * This method implements sophisticated caching behavior:
   * - Fresh prompts are returned immediately from cache
   * - Expired prompts are returned from cache while being refreshed in background
   * - Cache misses trigger immediate fetch with optional fallback support
   *
   * @param name - Name of the prompt to retrieve
   * @param options - Optional retrieval configuration
   * @returns Promise that resolves to the appropriate prompt client
   *
   * @example
   * ```typescript
   * // Get latest version with caching
   * const prompt = await langfuse.prompt.get("my-prompt");
   *
   * // Get specific version
   * const v2Prompt = await langfuse.prompt.get("my-prompt", {
   *   version: 2
   * });
   *
   * // Get with label filter
   * const prodPrompt = await langfuse.prompt.get("my-prompt", {
   *   label: "production"
   * });
   *
   * // Get with fallback
   * const promptWithFallback = await langfuse.prompt.get("my-prompt", {
   *   type: "text",
   *   fallback: "Hello {{name}}!"
   * });
   * ```
   */
  async get(name, options) {
    var _a;
    const cacheKey = this.cache.createKey({
      name,
      label: options == null ? void 0 : options.label,
      version: options == null ? void 0 : options.version
    });
    const cachedPrompt = this.cache.getIncludingExpired(cacheKey);
    if (!cachedPrompt || (options == null ? void 0 : options.cacheTtlSeconds) === 0) {
      try {
        return await this.fetchPromptAndUpdateCache({
          name,
          version: options == null ? void 0 : options.version,
          label: options == null ? void 0 : options.label,
          cacheTtlSeconds: options == null ? void 0 : options.cacheTtlSeconds,
          maxRetries: options == null ? void 0 : options.maxRetries,
          fetchTimeoutMs: options == null ? void 0 : options.fetchTimeoutMs
        });
      } catch (err) {
        if (options == null ? void 0 : options.fallback) {
          const sharedFallbackParams = {
            name,
            version: (_a = options == null ? void 0 : options.version) != null ? _a : 0,
            labels: options.label ? [options.label] : [],
            cacheTtlSeconds: options == null ? void 0 : options.cacheTtlSeconds,
            config: {},
            tags: []
          };
          if (options.type === "chat") {
            return new ChatPromptClient(
              {
                ...sharedFallbackParams,
                type: "chat",
                prompt: options.fallback.map((msg) => ({
                  ...msg,
                  type: "chatmessage" /* ChatMessage */
                }))
              },
              true
            );
          } else {
            return new TextPromptClient(
              {
                ...sharedFallbackParams,
                type: "text",
                prompt: options.fallback
              },
              true
            );
          }
        }
        throw err;
      }
    }
    if (cachedPrompt.isExpired) {
      if (!this.cache.isRefreshing(cacheKey)) {
        const refreshPromptPromise = this.fetchPromptAndUpdateCache({
          name,
          version: options == null ? void 0 : options.version,
          label: options == null ? void 0 : options.label,
          cacheTtlSeconds: options == null ? void 0 : options.cacheTtlSeconds,
          maxRetries: options == null ? void 0 : options.maxRetries,
          fetchTimeoutMs: options == null ? void 0 : options.fetchTimeoutMs
        }).catch(() => {
          this.logger.warn(
            `Failed to refresh prompt cache '${cacheKey}', stale cache will be used until next refresh succeeds.`
          );
        });
        this.cache.addRefreshingPromise(cacheKey, refreshPromptPromise);
      }
      return cachedPrompt.value;
    }
    return cachedPrompt.value;
  }
  async fetchPromptAndUpdateCache(params) {
    const cacheKey = this.cache.createKey(params);
    try {
      const {
        name,
        version,
        cacheTtlSeconds,
        label,
        maxRetries,
        fetchTimeoutMs
      } = params;
      const data = await this.apiClient.prompts.get(
        name,
        {
          version,
          label
        },
        {
          maxRetries,
          timeoutInSeconds: fetchTimeoutMs ? fetchTimeoutMs / 1e3 : void 0
        }
      );
      let prompt;
      if (data.type === "chat") {
        prompt = new ChatPromptClient(data);
      } else {
        prompt = new TextPromptClient(data);
      }
      this.cache.set(cacheKey, prompt, cacheTtlSeconds);
      return prompt;
    } catch (error) {
      this.logger.error(`Error fetching prompt '${cacheKey}':`, error);
      throw error;
    }
  }
};

// src/score/index.ts
import {
  getEnv,
  generateUUID,
  getGlobalLogger as getGlobalLogger5,
  LogLevel,
  safeSetTimeout
} from "@langfuse/core";
import { trace as trace2 } from "@opentelemetry/api";
var MAX_QUEUE_SIZE = 1e5;
var MAX_BATCH_SIZE = 100;
var ScoreManager = class {
  /**
   * Creates a new ScoreManager instance.
   *
   * @param params - Configuration object containing the API client
   * @internal
   */
  constructor(params) {
    this.eventQueue = [];
    this.flushPromise = null;
    this.flushTimer = null;
    this.apiClient = params.apiClient;
    const envFlushAtCount = getEnv("LANGFUSE_FLUSH_AT");
    const envFlushIntervalSeconds = getEnv("LANGFUSE_FLUSH_INTERVAL");
    this.flushAtCount = envFlushAtCount ? Number(envFlushAtCount) : 10;
    this.flushIntervalSeconds = envFlushIntervalSeconds ? Number(envFlushIntervalSeconds) : 1;
  }
  get logger() {
    return getGlobalLogger5();
  }
  /**
   * Creates a new score event and adds it to the processing queue.
   *
   * Scores are queued and sent in batches for efficiency. The score will be
   * automatically sent when the queue reaches the flush threshold or after
   * the flush interval expires.
   *
   * @param data - The score data to create
   *
   * @example
   * ```typescript
   * langfuse.score.create({
   *   name: "quality",
   *   value: 0.85,
   *   traceId: "trace-123",
   *   comment: "High quality response"
   * });
   * ```
   */
  create(data) {
    var _a, _b;
    const scoreData = {
      ...data,
      id: (_a = data.id) != null ? _a : generateUUID(),
      environment: (_b = data.environment) != null ? _b : getEnv("LANGFUSE_TRACING_ENVIRONMENT")
    };
    const scoreIngestionEvent = {
      id: generateUUID(),
      type: "score-create",
      timestamp: (/* @__PURE__ */ new Date()).toISOString(),
      body: scoreData
    };
    if (this.eventQueue.length >= MAX_QUEUE_SIZE) {
      this.logger.error(
        `Score queue is at max size ${MAX_QUEUE_SIZE}. Dropping score.`
      );
      return;
    }
    this.eventQueue.push(scoreIngestionEvent);
    if (this.logger.isLevelEnabled(LogLevel.DEBUG)) {
      this.logger.debug(
        "Added score event to queue:\n",
        JSON.stringify(scoreIngestionEvent, null, 2)
      );
    }
    if (this.eventQueue.length >= this.flushAtCount) {
      this.flushPromise = this.flush();
    } else if (!this.flushTimer) {
      this.flushTimer = safeSetTimeout(() => {
        this.flushPromise = this.flush();
      }, this.flushIntervalSeconds * 1e3);
    }
  }
  /**
   * Creates a score for a specific observation using its OpenTelemetry span.
   *
   * This method automatically extracts the trace ID and observation ID from
   * the provided span context.
   *
   * @param observation - Object containing the OpenTelemetry span
   * @param data - Score data (traceId and observationId will be auto-populated)
   *
   * @example
   * ```typescript
   * import { startSpan } from '@langfuse/tracing';
   *
   * const span = startSpan({ name: "my-operation" });
   * langfuse.score.observation(
   *   { otelSpan: span },
   *   { name: "accuracy", value: 0.92 }
   * );
   * ```
   */
  observation(observation, data) {
    const { spanId, traceId } = observation.otelSpan.spanContext();
    this.create({
      ...data,
      traceId,
      observationId: spanId
    });
  }
  /**
   * Creates a score for a trace using an OpenTelemetry span.
   *
   * This method automatically extracts the trace ID from the provided
   * span context and creates a trace-level score.
   *
   * @param observation - Object containing the OpenTelemetry span
   * @param data - Score data (traceId will be auto-populated)
   *
   * @example
   * ```typescript
   * import { startSpan } from '@langfuse/tracing';
   *
   * const span = startSpan({ name: "my-operation" });
   * langfuse.score.trace(
   *   { otelSpan: span },
   *   { name: "overall_quality", value: 0.88 }
   * );
   * ```
   */
  trace(observation, data) {
    const { traceId } = observation.otelSpan.spanContext();
    this.create({
      ...data,
      traceId
    });
  }
  /**
   * Creates a score for the currently active observation.
   *
   * This method automatically detects the active OpenTelemetry span and
   * creates an observation-level score. If no active span is found,
   * a warning is logged and the operation is skipped.
   *
   * @param data - Score data (traceId and observationId will be auto-populated)
   *
   * @example
   * ```typescript
   * import { startActiveSpan } from '@langfuse/tracing';
   *
   * startActiveSpan({ name: "my-operation" }, (span) => {
   *   // Inside the active span
   *   langfuse.score.activeObservation({
   *     name: "relevance",
   *     value: 0.95
   *   });
   * });
   * ```
   */
  activeObservation(data) {
    const currentOtelSpan = trace2.getActiveSpan();
    if (!currentOtelSpan) {
      this.logger.warn("No active span in context to score.");
      return;
    }
    const { spanId, traceId } = currentOtelSpan.spanContext();
    this.create({
      ...data,
      traceId,
      observationId: spanId
    });
  }
  /**
   * Creates a score for the currently active trace.
   *
   * This method automatically detects the active OpenTelemetry span and
   * creates a trace-level score. If no active span is found,
   * a warning is logged and the operation is skipped.
   *
   * @param data - Score data (traceId will be auto-populated)
   *
   * @example
   * ```typescript
   * import { startActiveSpan } from '@langfuse/tracing';
   *
   * startActiveSpan({ name: "my-operation" }, (span) => {
   *   // Inside the active span
   *   langfuse.score.activeTrace({
   *     name: "user_satisfaction",
   *     value: 4,
   *     comment: "User rated 4 out of 5 stars"
   *   });
   * });
   * ```
   */
  activeTrace(data) {
    const currentOtelSpan = trace2.getActiveSpan();
    if (!currentOtelSpan) {
      this.logger.warn("No active span in context to score trace.");
      return;
    }
    const { traceId } = currentOtelSpan.spanContext();
    this.create({
      ...data,
      traceId
    });
  }
  async handleFlush() {
    try {
      if (this.flushTimer) {
        clearTimeout(this.flushTimer);
        this.flushTimer = null;
      }
      const promises = [];
      while (this.eventQueue.length > 0) {
        const batch = this.eventQueue.splice(0, MAX_BATCH_SIZE);
        promises.push(
          this.apiClient.ingestion.batch({ batch }).then((res) => {
            var _a;
            if (((_a = res.errors) == null ? void 0 : _a.length) > 0) {
              this.logger.error("Error ingesting scores:", res.errors);
            }
          }).catch((err) => {
            this.logger.error("Failed to export score batch:", err);
          })
        );
      }
      await Promise.all(promises);
    } catch (err) {
      this.logger.error("Error flushing Score Manager: ", err);
    } finally {
      this.flushPromise = null;
    }
  }
  /**
   * Flushes all pending score events to the Langfuse API.
   *
   * This method ensures all queued scores are sent immediately rather than
   * waiting for the automatic flush interval or batch size threshold.
   *
   * @returns Promise that resolves when all pending scores have been sent
   *
   * @example
   * ```typescript
   * langfuse.score.create({ name: "quality", value: 0.8 });
   * await langfuse.score.flush(); // Ensures the score is sent immediately
   * ```
   */
  async flush() {
    var _a;
    return (_a = this.flushPromise) != null ? _a : this.handleFlush();
  }
  /**
   * Gracefully shuts down the score manager by flushing all pending scores.
   *
   * This method should be called before your application exits to ensure
   * all score data is sent to Langfuse.
   *
   * @returns Promise that resolves when shutdown is complete
   *
   * @example
   * ```typescript
   * // Before application exit
   * await langfuse.score.shutdown();
   * ```
   */
  async shutdown() {
    await this.flush();
  }
};

// src/LangfuseClient.ts
var LangfuseClient = class {
  /**
   * Creates a new LangfuseClient instance.
   *
   * @param params - Configuration parameters. If not provided, will use environment variables.
   *
   * @throws Will log warnings if required credentials are not provided
   *
   * @example
   * ```typescript
   * // With explicit configuration
   * const client = new LangfuseClient({
   *   publicKey: "pk_...",
   *   secretKey: "sk_...",
   *   baseUrl: "https://your-instance.langfuse.com"
   * });
   *
   * // Using environment variables
   * const client = new LangfuseClient();
   * ```
   */
  constructor(params) {
    this.projectId = null;
    var _a, _b, _c, _d, _e, _f, _g;
    const logger = getGlobalLogger6();
    const publicKey = (_a = params == null ? void 0 : params.publicKey) != null ? _a : getEnv2("LANGFUSE_PUBLIC_KEY");
    const secretKey = (_b = params == null ? void 0 : params.secretKey) != null ? _b : getEnv2("LANGFUSE_SECRET_KEY");
    this.baseUrl = (_e = (_d = (_c = params == null ? void 0 : params.baseUrl) != null ? _c : getEnv2("LANGFUSE_BASE_URL")) != null ? _d : getEnv2("LANGFUSE_BASEURL")) != null ? _e : (
      // legacy v2
      "https://cloud.langfuse.com"
    );
    if (!publicKey) {
      logger.warn(
        "No public key provided in constructor or as LANGFUSE_PUBLIC_KEY env var. Client operations will fail."
      );
    }
    if (!secretKey) {
      logger.warn(
        "No secret key provided in constructor or as LANGFUSE_SECRET_KEY env var. Client operations will fail."
      );
    }
    const timeoutSeconds = (_g = params == null ? void 0 : params.timeout) != null ? _g : Number((_f = getEnv2("LANGFUSE_TIMEOUT")) != null ? _f : 5);
    this.api = new LangfuseAPIClient4({
      baseUrl: this.baseUrl,
      username: publicKey,
      password: secretKey,
      xLangfusePublicKey: publicKey,
      xLangfuseSdkVersion: LANGFUSE_SDK_VERSION,
      xLangfuseSdkName: "javascript",
      environment: "",
      // noop as baseUrl is set
      headers: params == null ? void 0 : params.additionalHeaders
    });
    logger.debug("Initialized LangfuseClient with params:", {
      publicKey,
      baseUrl: this.baseUrl,
      timeoutSeconds
    });
    this.prompt = new PromptManager({ apiClient: this.api });
    this.dataset = new DatasetManager({ langfuseClient: this });
    this.score = new ScoreManager({ apiClient: this.api });
    this.media = new MediaManager({ apiClient: this.api });
    this.experiment = new ExperimentManager({ langfuseClient: this });
    this.getPrompt = this.prompt.get.bind(this.prompt);
    this.createPrompt = this.prompt.create.bind(this.prompt);
    this.updatePrompt = this.prompt.update.bind(this.prompt);
    this.getDataset = this.dataset.get;
    this.fetchTrace = this.api.trace.get;
    this.fetchTraces = this.api.trace.list;
    this.fetchObservation = this.api.legacy.observationsV1.get;
    this.fetchObservations = this.api.observations.getMany;
    this.fetchSessions = this.api.sessions.get;
    this.getDatasetRun = this.api.datasets.getRun;
    this.getDatasetRuns = this.api.datasets.getRuns;
    this.createDataset = this.api.datasets.create;
    this.getDatasetItem = this.api.datasetItems.get;
    this.createDatasetItem = this.api.datasetItems.create;
    this.fetchMedia = this.api.media.get;
    this.resolveMediaReferences = this.media.resolveReferences;
  }
  /**
   * Flushes any pending score events to the Langfuse API.
   *
   * This method ensures all queued scores are sent immediately rather than
   * waiting for the automatic flush interval or batch size threshold.
   *
   * @returns Promise that resolves when all pending scores have been sent
   *
   * @example
   * ```typescript
   * langfuse.score.create({ name: "quality", value: 0.8 });
   * await langfuse.flush(); // Ensures the score is sent immediately
   * ```
   */
  async flush() {
    return this.score.flush();
  }
  /**
   * Gracefully shuts down the client by flushing all pending data.
   *
   * This method should be called before your application exits to ensure
   * all data is sent to Langfuse.
   *
   * @returns Promise that resolves when shutdown is complete
   *
   * @example
   * ```typescript
   * // Before application exit
   * await langfuse.shutdown();
   * ```
   */
  async shutdown() {
    return this.score.shutdown();
  }
  /**
   * Generates a URL to view a specific trace in the Langfuse UI.
   *
   * @param traceId - The ID of the trace to generate a URL for
   * @returns Promise that resolves to the trace URL
   *
   * @example
   * ```typescript
   * const traceId = "trace-123";
   * const url = await langfuse.getTraceUrl(traceId);
   * console.log(`View trace at: ${url}`);
   * ```
   */
  async getTraceUrl(traceId) {
    let projectId = this.projectId;
    if (!projectId) {
      projectId = (await this.api.projects.get()).data[0].id;
      this.projectId = projectId;
    }
    const traceUrl = `${this.baseUrl}/project/${projectId}/traces/${traceId}`;
    return traceUrl;
  }
};

// src/experiment/adapters.ts
function createEvaluatorFromAutoevals(autoevalEvaluator, params) {
  const langfuseEvaluator = async (langfuseEvaluatorParams) => {
    var _a;
    const score = await autoevalEvaluator({
      ...params != null ? params : {},
      input: langfuseEvaluatorParams.input,
      output: langfuseEvaluatorParams.output,
      expected: langfuseEvaluatorParams.expectedOutput
    });
    return {
      name: score.name,
      value: (_a = score.score) != null ? _a : 0,
      metadata: score.metadata
    };
  };
  return langfuseEvaluator;
}
export {
  ChatMessageType,
  ChatPromptClient,
  DatasetManager,
  ExperimentManager,
  LangfuseClient,
  MediaManager,
  PromptManager,
  ScoreManager,
  TextPromptClient,
  createEvaluatorFromAutoevals
};
